### PR TITLE
Decoder

### PR DESCRIPTION
**Pull request by lorelupo**
Mi sa che era github, sono rimasto connesso a TIM e il problema sembra risolto

### CODE REVIEW

File: src/parameters.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        self.out_channels = 3  # to check
+        self.kernel_sizes = [1, 2, 3]  # to check
+        self.embedding_size = embedding_size
+        self.hidden_size = 5
+        self.dropout = 0.5
+        self.learning_rate = 0.0001
+        self.betas = (0.5, 0.999)
+
+
+class Autoencoder(object):
```
Comment (mberaha) [inline src/parameters.py]: AutoencoderParams

File: src/parameters.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+class Discriminator(object):
```
Comment (mberaha) [inline src/parameters.py]: Qui può esserci un problema di nomi fra la classe Discriminator e questa.
Meglio DiscriminatorParams

File: src/parameters.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+class Params(object):
+    in_memory = True
+    max_length = 20
+    embedding_size = 200
+    dim_y = 200
+    dim_z = 500
+    batch_size = 128
+    temperature = 0.001
+    lambda_GAN = 1
+    autoencoder = Autoencoder(embedding_size, dim_y, dim_z)
```
Comment (mberaha) [inline src/parameters.py]: autoencoder = AutoencoderParams(...)

File: src/parameters.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    discriminator = Discriminator(embedding_size)
```
Comment (mberaha) [inline src/parameters.py]: discriminator = DiscriminatorParams()

File: src/style_transfer.py [hunk 5]
```diff
@@ -1,4 +1,246 @@
...
+        self.autoencoder_optimizer = optim.Adam(
+            [{'params': self.encoder.parameters()},
+             {'params': self.generator.parameters()},
+             {'params': self.labelsTransform.parameters()},
+             {'params': self.vocabulary.embeddings.parameters()},
+             {'params': self.hiddenToVocab.parameters()}],
+            lr=params.autoencoder.learning_rate,
+            betas=params.autoencoder.betas)
+        self.discriminator0_optimizer = optim.Adam(
+            self.discriminator.parameters(),
```
Comment (mberaha) [inline src/style_transfer.py]: self.discriminators[0].parameters()

File: src/style_transfer.py [hunk 6]
```diff
@@ -1,4 +1,246 @@
...
+            lr=params.discriminator.learning_rate,
+            betas=params.discriminator.betas)
+        self.discriminator1_optimizer = optim.Adam(
+            self.discriminator.parameters(),
```
Comment (mberaha) [inline src/style_transfer.py]: self.discriminators[1].parameters()

File: src/style_transfer.py [hunk 7]
```diff
@@ -1,4 +1,246 @@
...
+        self.generator = Rnn(
+            params.autoencoder.input_size,
+            params.autoencoder.hidden_size,
+            params.autoencoder.num_layers,
+            device)
+
+        # instantiating linear networks for hidden transformations
+        self.labelsTransform = torch.nn.Linear(1, params.dim_y)
+        self.hiddenToVocab = torch.nn.Linear(
+            params.hidden_size, self.vocabulary.vocabSize + 1)
```
Comment (mberaha) [inline src/style_transfer.py]: params.autoencoder.hidden_size

File: src/style_transfer.py [hunk 8]
```diff
@@ -1,4 +1,246 @@
...
+        loss_g = self.adv_loss_criterion(d_fake, label)
+        return loss_d, loss_g
+
+    def _zeroGradients(self):
+        self.autoencoder_optimizer.zero_grad()
+        self.discriminator0_optimizer.zero_grad()
+        self.discriminator1_optimizer.zero_grad()
+
+    def _runSentence(self, encoder_input, generator_input, label, target):
+
```
Comment (mberaha) [inline src/style_transfer.py]: rimuovere linea bianca

**Review by mberaha — APPROVED** [review]
PR ottima, solo qualche modifica qui e la ma nulla di sostanziale. Modifica e poi fail il merge quando vuoi!!

File: src/generate_batches.py [hunk 9]
```diff
@@ -30,21 +30,37 @@ def yieldBatchesFromFiles(files, batchsize):
 
 def loadFilesAndGenerateBatches(files, batchsize, shuffleFiles=True):
     inputs = []
-    labels = []
+    lenLines = []
     for label, fileName in enumerate(files):
```
Comment (WolfLo) [inline src/generate_batches.py]: come usi la variabile label in questo for loop?

Comment (lorelupo) [inline src/generate_batches.py:34]: comeusi la variabile label dentro questo for loop?
